src/experiments/settings/api.js

```diff
@@ -0,0 +1,164 @@
+"use strict";
+
+
+/* global Components, ExtensionAPI */
+let Cu2 = Components.utils;
+Cu2.import("resource://gre/modules/Services.jsm");
+Cu2.import("resource://gre/modules/ExtensionSettingsStore.jsm");
+Cu2.import("resource://gre/modules/ExtensionPreferencesManager.jsm");
+
+
+// TODO file scope issue on experiments that join extension contexts causing redeclaration issues.
+
+const prefManager = {
+  setPrefs(prefs) {
+    prefs.forEach((pref) => {
+      this.setPref(pref.name, pref.value, pref.type);
+    });
+  },
+  setPref(name, value, type) {
+    if (value === null) {
+      Services.prefs.clearUserPref(name);
+    }
+    /* As prefs are hidden we can't use Services.prefs.getPrefType */
+    switch (type) {
+      case "string":
+        return Services.prefs.setCharPref(name, value);
+      case "int":
+        return Services.prefs.setIntPref(name, value);
+      case "bool":
+        return Services.prefs.setBoolPref(name, value);
+      default:
+        throw new Error("Unknown type");
+    }
+  },
+
+  getUserPrefs(prefNames) {
+    let prefs = {};
+    prefNames.forEach((prefName) => {prefs[prefName] = this.getUserPref(prefName)});
+    return prefs;
+  },
+  getUserPref(name, value) {
+    if (!Services.prefs.prefHasUserValue(name)) {
+      return null;
+    }
+    let type = Services.prefs.getPrefType(name);
+    switch (type) {
+      case Services.prefs.PREF_STRING:
+        return Services.prefs.getCharPref(name, value);
+      case Services.prefs.PREF_INT:
+        return Services.prefs.getIntPref(name, value);
+      case Services.prefs.PREF_BOOL:
+        return Services.prefs.getBoolPref(name, value);
+      default:
+        throw new Error("Unknown type");
+    }
+  }
+}
+const SETTING_TYPE = "setting_config";
+const settingManager = {
+  async init() {
+    await ExtensionSettingsStore.initialize();
+  },
+  async add(id, settingConfig) {
+    await this.init();
+    await ExtensionSettingsStore.addSetting(id, SETTING_TYPE, `settingConfig_${settingConfig.name}`, settingConfig);
```

Comment:
Having looked into this for a while I get the feeling you're just duplicating efforts with ExtensionPreferencesManager by maintaining a separate ExtensionSettingsStore setting.

Can you try just ripping out ExtensionSettingsStore usage here and using `ExtensionPreferencesManager.getSetting()` for `settings.get()` instead?

https://searchfox.org/mozilla-central/rev/d4b9e50875ad7e5d20f2fee6a53418315f6dfcc0/toolkit/components/extensions/ExtensionPreferencesManager.jsm#314

If that works this code would get much simpler...